### PR TITLE
refactor: introduce PersistentObjectFactory and deprecate PersistentProxyObjectFactory with final

### PR DESCRIPTION
**Pull request by nikophil**

### CODE REVIEW

File: src/Factory.php [hunk 1]
```diff
@@ -87,11 +88,23 @@ public function __call(string $name, array $arguments): array
...
+    final public function create(
+        array|callable $attributes = [],
+        /**
+         * @deprecated
+         * @internal
+         */
+        bool $noProxy = false
+    ): object
     {
+        if (\count(func_get_args()) === 2 && !str_starts_with(debug_backtrace(options: \DEBUG_BACKTRACE_IGNORE_ARGS, limit: 2)[1]['class'] ?? '', 'Zenstruck\Foundry')) {
```
Comment (kbond) [inline src/Factory.php]: Cool solution!

File: src/Factory.php [hunk 2]
```diff
@@ -483,4 +497,10 @@ private function withCascadePersist(): static
 
         return $cloned;
     }
+
+    /** @interal */
```
Comment (kbond) [inline src/Factory.php]:
```suggestion
    /**
     * @internal
     */
```

File: src/Persistence/PersistentProxyObjectFactory.php [hunk 3]
```diff
@@ -14,22 +14,28 @@
...
+abstract class PersistentProxyObjectFactory extends PersistentObjectFactory
 {
-    public function __construct()
+    final public static function new(array|callable|string $defaultAttributes = [], string ...$states): static
     {
-        parent::__construct(static::class(), calledInternally: true);
+        if ((new \ReflectionClass(static::class()))->isFinal()) {
+            trigger_deprecation(
+                'zenstruck\foundry', '1.37.0',
+                \sprintf('Using a proxy factory with a final class is deprecated and will throw an error in Foundry 2.0. Use "%s" instead, or unfinalize "%s" class.', PersistentProxyObjectFactory::class, static::class())
```
Comment (kbond) [inline src/Persistence/PersistentProxyObjectFactory.php]: nice call here!

File: src/Factory.php (hunk 1 above)
Comment (nikophil) [reply src/Factory.php]: In one of the previous BC PR, I added few parameters called `calledInternally`, I think I'll drop them in favor of this mechanism

File: psalm.xml [hunk 4]
```diff
@@ -12,4 +12,9 @@
         <file name="tests/Fixtures/Maker/expected/can_create_factory_for_entity_with_repository_with_data_set_psalm.php" />
         <file name="tests/Fixtures/Maker/expected/can_create_factory_with_static_analysis_annotations_with_data_set_psalm.php" />
     </projectFiles>
+
+    <issueHandlers>
+        <ImplementedReturnTypeMismatch errorLevel="suppress" />
+        <MethodSignatureMismatch errorLevel="suppress" />
+    </issueHandlers>
```
Comment (nikophil) [inline psalm.xml:19]: this is not ideal, but... psalm analysis passes :innocent:

File: src/Factory.php [hunk 5]
```diff
@@ -87,11 +88,23 @@ public function __call(string $name, array $arguments): array
...
+     * @psalm-suppress ImplementedReturnTypeMismatch
      */
-    final public function create(array|callable $attributes = []): Proxy
+    final public function create(
+        array|callable $attributes = [],
+        /**
+         * @deprecated
+         * @internal
+         */
+        bool $noProxy = false
```
Comment (nikophil) [inline src/Factory.php:101]: we could prevent adding this parameter, and create a method:
```php
public function withProxy(): bool
{
    return $this instanceof PersistentObjectProxyFactory::class;
}
```
but this way, we cannot use the phpstan trick `@phpstan-return ($noProxy is true ? TObject: Proxy<TObject>)`
and it breaks the type system in multiple places :shrug: 

So I'd prefer to use this technique which is far from perfect but which seems acceptable for  a BC layer, WDYT?

File: phpstan-baseline.neon [hunk 6]
```diff
@@ -1,5 +1,10 @@
 parameters:
 	ignoreErrors:
+		-
+			message: "#^Should not use node with type \"Expr_Eval\", please change the code\\.$#"
+			count: 1
+			path: src/AnonymousFactoryGenerator.php
+
```
Comment (nikophil) [inline phpstan-baseline.neon:7]: this error comes from `eval()` in `AnonymousFactoryGenerator`

**Review by kbond — APPROVED** [review]

**Pull request merged**